File: ssp.py
from typing import List
from queue import Queue


class Solution:

    def isValid(self, x: int, y: int, n: int, m: int) -> bool:
        return x >= 0 and x <= n-1 and y >= 0 and y <= m-1

    def shortestPath(self, grid: List[List[int]], k: int) -> int:

        # find the dimensions of the grid
        n = len(grid)
        m = len(grid[0])

        # Direction vector
        dir = [0, 1, 0, -1, 0]

        steps = 0

        # lives array to optimize the use of obstacle removal power
        lives = [[-1 for j in range(m)] for i in range(n)]

        # Declare a queue to push the elements
        qu = Queue()

        qu.put([0, 0, k])

        while not qu.empty():

            size = qu.qsize()

            for i in range(size):

                row, col, currLives = qu.get()

                if row == n-1 and col == m-1:
                    return steps

                if grid[row][col] == 1:
                    currLives -= 1

                for i in range(4):
                    newRow = row + dir[i]
                    newCol = col + dir[i+1]

                    if self.isValid(newRow, newCol, n, m) and lives[newRow][newCol] < currLives:
                        lives[newRow][newCol] = currLives
                        qu.put([newRow, newCol, currLives])

            steps += 1

        return -1

File: ssp.java
import java.util.*;

class Solution {

    public boolean isValid(int x, int y, int n, int m) {
        return x >= 0 && x <= n-1 && y >= 0 && y <= m-1;
    }

    public int shortestPath(List<List<Integer>> grid, int k) {

        // find the dimensions of the grid
        int n = grid.size();
        int m = grid.get(0).size();

        // Direction vector
        List<Integer> dir = Arrays.asList(0, 1, 0, -1, 0);

        int steps = 0;

        // Lives array to optimize the use of obstacle removal power
        int[][] lives = new int[n][m];
        for (int i = 0; i < n; i++) {
            Arrays.fill(lives[i], -1);
        }

        // Declare a queue to push the elements
        Queue<List<Integer>> qu = new LinkedList<>();
        qu.offer(Arrays.asList(0, 0, k));

        while (!qu.isEmpty()) {

            int size = qu.size();

            for (int i = 0; i < size; i++) {

                int row = qu.peek().get(0);
                int col = qu.peek().get(1);
                int currLives = qu.peek().get(2);

                qu.poll();

                if (row == n-1 && col == m-1) {
                    return steps;
                }

                if (grid.get(row).get(col) == 1) {
                    currLives--;
                }

                for (int j = 0; j < 4; j++) {
                    int newRow = row + dir.get(j);
                    int newCol = col + dir.get(j+1);

                    if (isValid(newRow, newCol, n, m) && lives[newRow][newCol] < currLives) {
                        lives[newRow][newCol] = currLives;
                        qu.offer(Arrays.asList(newRow, newCol, currLives));
                    }
                }
            }

            steps++;
        }

        return -1;
    }
}

            



